File: packages/sdk-ts/src/admin/audit.ts
// SPDX-License-Identifier: Apache-2.0

// suite.admin.audit.* — audit log search.
//
// Endpoint (mirroring `apps/dashboard/src/lib/api.ts`):
//
//   GET /api/v1/admin/audit?tenant=&action=&from=&to=&limit=&offset=

import { request, toCamel, type HttpOptions } from "../_http.js"
import { AuditListSchema, type AuditList } from "./_models.js"

export interface ListAuditOptions extends HttpOptions {
  tenant?: string
  action?: string
  /** RFC3339 timestamp or Date. */
  from?: string | Date
  /** RFC3339 timestamp or Date. */
  to?: string | Date
  limit?: number
  offset?: number
}

/** List audit log entries with optional filters. */
export async function list(opts: ListAuditOptions = {}): Promise<AuditList> {
  const {
    tenant,
    action,
    from,
    to,
    limit = 100,
    offset = 0,
    ...http
  } = opts
  if (limit <= 0) throw new Error("limit must be a positive integer")
  if (offset < 0) throw new Error("offset must be non-negative")
  const qs = new URLSearchParams()
  if (tenant !== undefined && tenant !== "") qs.set("tenant", tenant)
  if (action !== undefined && action !== "") qs.set("action", action)
  if (from !== undefined) qs.set("from", toIso(from))
  if (to !== undefined) qs.set("to", toIso(to))
  qs.set("limit", String(limit))
  qs.set("offset", String(offset))
  const raw = await request<unknown>("GET", `/admin/audit?${qs.toString()}`, null, http)
  return AuditListSchema.parse(toCamel(raw))
}

function toIso(value: string | Date): string {
  return value instanceof Date ? value.toISOString() : value
}

/** Namespace object — the shape `suite.admin.audit` is built from. */
export const audit = {
  list,
} as const

File: packages/sdk-py/af_stack/admin/audit.py
# SPDX-License-Identifier: Apache-2.0

"""``suite.admin.audit.*`` — audit log search.

Maps to:

==================================  =========================================
SDK call                            Endpoint
==================================  =========================================
:func:`list`                        ``GET /api/v1/admin/audit``
==================================  =========================================

The endpoint accepts free-form filters via query params; results are
ordered ``occurred_at desc`` and paginated.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from .. import _http
from ._models import AuditList


async def list(  # noqa: A001 — mirrors the JS ``audit.list()`` ergonomics
    *,
    tenant: str | None = None,
    action: str | None = None,
    from_: str | datetime | None = None,
    to: str | datetime | None = None,
    limit: int = 100,
    offset: int = 0,
) -> AuditList:
    """List audit log entries with optional filters.

    ``from_`` and ``to`` are RFC3339 timestamps (strings) or
    :class:`datetime` objects. ``from_`` uses a trailing underscore
    because ``from`` is a Python keyword; the wire param is plain
    ``from``.
    """
    if limit <= 0:
        raise ValueError("limit must be a positive integer")
    if offset < 0:
        raise ValueError("offset must be non-negative")
    params: dict[str, Any] = {"limit": limit, "offset": offset}
    if tenant:
        params["tenant"] = tenant
    if action:
        params["action"] = action
    if from_ is not None:
        params["from"] = _to_iso(from_)
    if to is not None:
        params["to"] = _to_iso(to)
    body = await _http.request_json("GET", "/admin/audit", params=params)
    return AuditList.model_validate(body or {"entries": [], "total": 0, "has_more": False})


def _to_iso(value: str | datetime) -> str:
    """Serialise a :class:`datetime` to RFC3339, leaving strings alone.

    Naive datetimes are treated as UTC for predictability — same
    convention used by :mod:`af_stack.jobs`.
    """
    if isinstance(value, datetime):
        iso = value.isoformat()
        if value.tzinfo is None:
            return iso + "Z"
        return iso
    return value


__all__ = ["list"]